pomegranate/callbacks.py

```diff
@@ -0,0 +1,97 @@
+
+class Callback(object):
+    """A base Callback class."""
+
+    def __init__(self):
+        self.model = None
+        self.params = {}
+
+    def set_model(self, model):
+        self.model = model
+
+    def set_params(self, params):
+        self.params = params
+
+    def on_train_begin(self):
+        pass
+
+    def on_train_end(self):
+        pass
+
+    def on_iteration_begin(self, iteration):
+        pass
+
+    def on_iteration_end(self, iteration, logged_vars=None):
+        pass
+
+
+class History(Callback):
+    """A simple History callback. Stores the history of the whole training."""
+
+    def on_train_begin(self):
+        self.history = {}
+        self.iterations = []
+
+    def on_iteration_end(self, iteration, logged_vars=None):
+        self.iterations.append(iteration)
+        for var, val in logged_vars.items():
+            self.history.setdefault(var, []).append(val)
+
+
+class ParamsTerminator(Callback):
```

Comment:
The point of this is to replace the use of a while loop? I think that loop should remain and the callbacks are additional things that you'd like to do. Otherwise the termination conditions need to be included in every callback you write. Perhaps we could have a property from the callback `continue_training` that defaults to true, and if it ever defaults to False then it stops training, overriding the other conditions?